### PR TITLE
Refactor involved projects

### PR DESCRIPTION
**Pull request by ChrisBr**

### CODE REVIEW

File: src/api/app/models/user.rb [hunk 1]
```diff
@@ -696,16 +700,7 @@ def delete!
...
-    role = Role.hashed['maintainer']
-
-    ### all projects where user is maintainer
-    projects = relationships.projects.where(role_id: role.id).pluck(:project_id)
-
-    # all projects where user is maintainer via a group
-    projects += Relationship.projects.where(role_id: role.id).joins(:groups_users).where(groups_users: { user_id: id }).pluck(:project_id)
-
-    projects.uniq
+    (direct_maintained_projects.pluck(:project_id) + group_maintained_projects.pluck(:project_id)).uniq
```
Comment (mschnitzer) [inline src/api/app/models/user.rb]: Is it possible to use a scope instead?

Comment (ChrisBr) [inline src/api/app/models/user.rb]: I guess not. Scopes are class methods, but I need this method on an instance (User.involved_projects_ids vs. User.first.involved_projects_ids). Furthermore these are two separated queries (direct vs group).

File: src/api/app/models/user.rb [hunk 2]
```diff
@@ -32,6 +32,10 @@ class User < ApplicationRecord
   has_many :groups_users, inverse_of: :user
   has_many :roles_users, inverse_of: :user
   has_many :relationships, inverse_of: :user, dependent: :destroy
+  has_many :direct_maintained_project_relationships, -> { projects.where(role_id: Role.hashed['maintainer'].id) }, class_name: 'Relationship'
+  has_many :direct_maintained_projects, through: :direct_maintained_project_relationships, source: :project
+  has_many :group_maintained_project_relationships, -> { projects.where(role_id: Role.hashed['maintainer'].id).includes(:project, :groups_users) }, through: :groups, source: :relationships
```
Comment (bgeuken) [inline src/api/app/models/user.rb]: IIRC you could shorten this to ```where(role: Role.hashed['maintainer'] ```

File: src/api/app/models/user.rb [hunk 3]
```diff
@@ -32,6 +32,10 @@ class User < ApplicationRecord
   has_many :groups_users, inverse_of: :user
   has_many :roles_users, inverse_of: :user
   has_many :relationships, inverse_of: :user, dependent: :destroy
+  has_many :direct_maintained_project_relationships, -> { projects.where(role_id: Role.hashed['maintainer'].id) }, class_name: 'Relationship'
```
Comment (bgeuken) [inline src/api/app/models/user.rb]: projects doesn't seem to be defined anywhere.

Comment (ChrisBr) [inline src/api/app/models/user.rb]: There is a scope in Relationships
``
  scope :projects, -> { where("project_id is not null") }
``

Comment (bgeuken) [inline src/api/app/models/user.rb]: Ah, I see.

File: src/api/app/models/user.rb (hunk 2 above)
Comment (ChrisBr) [inline src/api/app/models/user.rb]: Same with the scope

File: src/api/app/models/user.rb (hunk 3 above)
Comment (bgeuken) [inline src/api/app/models/user.rb]: Could we define a scope for the relationship mode (where role is maintainer) and then call ``` user.relationships.maintained``` ?

Seems so:
```
rb(main):013:0> User.last.relationships.joins(:role).where('roles.title="maintainer"')
=> #<ActiveRecord::AssociationRelation [#<Relationship id: 5, package_id: nil, project_id: 5, role_id: 2, user_id: 7, group_id: nil>]>
```

```
User.last.direct_maintained_project_relationships
=> #<ActiveRecord::Associations::CollectionProxy [#<Relationship id: 5, package_id: nil, project_id: 5, role_id: 2, user_id: 7, group_id: nil>]>
```

File: src/api/app/models/user.rb (hunk 2 above)
Comment (bgeuken) [inline src/api/app/models/user.rb]: > ChrisBr 4 minutes ago Member
> Same with the scope

What do you mean?

File: src/api/app/models/user.rb (hunk 3 above)
Comment (Ana06) [inline src/api/app/models/user.rb]: you can remove the `id`:

`projects.where(role: Role.hashed['maintainer']) }, class_name: 'Relationship'`

Comment (Ana06) [inline src/api/app/models/user.rb]: and why using `class_name` instead of `through`?

File: src/api/app/models/user.rb (hunk 1 above)
Comment (Ana06) [inline src/api/app/models/user.rb]: what about using `||` instead of `+`, then you don't need `uniq`:

`direct_maintained_projects.pluck(:project_id) || group_maintained_projects.pluck(:project_id)`

File: src/api/app/models/user.rb (hunk 3 above)
Comment (bgeuken) [inline src/api/app/models/user.rb]: I tend to think that instead of additional has_many relations we might be able to use existing relations and scopes

File: src/api/app/models/user.rb (hunk 1 above)
Comment (ChrisBr) [inline src/api/app/models/user.rb]: You mean one pipe ``|``, not two ``||`` ;)

I did a short benchmark which showed that it is actually slower:
```
Benchmark.measure { (1..100).each { u.involved_projects_ids } } 
uniq: 2.180000   0.440000   2.620000 (  8.730327)
pipe: 2.610000   0.670000   3.280000 (  9.440138)
```

**Comment by ChrisBr** [comment]
@Ana06 @bgeuken ok, I got rid of most of the additional relations and used the already existing relations and scopes!

**Comment by bgeuken** [comment]
Looks great

**Comment by ChrisBr** [comment]
@hennevogel how can this decrease by 24% ?

File: src/api/app/models/user.rb [hunk 4]
```diff
@@ -696,16 +697,7 @@ def delete!
...
-    role = Role.hashed['maintainer']
-
-    ### all projects where user is maintainer
-    projects = relationships.projects.where(role_id: role.id).pluck(:project_id)
-
-    # all projects where user is maintainer via a group
-    projects += Relationship.projects.where(role_id: role.id).joins(:groups_users).where(groups_users: { user_id: id }).pluck(:project_id)
-
-    projects.uniq
+    (relationships.projects.maintainers.pluck(:project_id) + group_relationships.projects.maintainers.pluck(:project_id)).uniq
```
Comment (Ana06) [inline src/api/app/models/user.rb]: what about using `|` instead of `+`, then you don't need `uniq`:

`direct_maintained_projects.pluck(:project_id) | group_maintained_projects.pluck(:project_id)`

**Comment by hennevogel** [comment]
@ChrisBr there is no successful build for the parent commit to compare to...

File: src/api/app/models/user.rb (hunk 4 above)
Comment (ChrisBr) [inline src/api/app/models/user.rb]: After some discussions: I still think this is more explicit and and on my machine it was even slightly faster (even that stackoverflow says that they almost equal)

**Comment by ChrisBr** [comment]
@hennevogel so can I merge then?
